### PR TITLE
Add tooltip to bar tile and web component

### PR DESCRIPTION
**Pull request by juliawu**
Adds a tooltip when hovering over the bars or lollipop circles in a bar tile. 

Specific changes:
* Add tooltip functions to `draw_bar.ts`
* Pipe through observation dates from data fetch to draw function to allow dates to be shown in the tooltips
* Move line chart specific tooltip function from utils into `draw_line.ts`
* Make Tooltip IDs unique for both bar and line charts
* 
* Enable changes in explore interface.

Note that:
* For stacked bar charts, only the segment highlighted has data shown. Calculating the total sum of the stack and showing that should come in a future PR
* Showing data for the entire group in grouped bar charts should come in a future PR
* For lollipops, only hovering over the circles works. Grouping the circles and stems together should come in a future PR.

Screenshots:
Screenshot 2023-08-21 at 9 27 06 AM

Screenshot 2023-08-21 at 9 27 26 AM

Screenshot 2023-08-21 at 9 27 42 AM

### CODE REVIEW

**Comment by pradh** [comment]
This is so useful, thanks for adding!

File: static/js/chart/draw_bar.ts [hunk 1]
```diff
@@ -496,16 +699,25 @@ function drawHorizontalStackedBars(
       .append("g")
       .attr("fill", (d) => colorFn(d.key))
       .selectAll("circle")
-      .data((d) => d.map((dp) => ({ key: d.key, ...dp })))
+      .data((d) =>
+        d.map((dp) => ({
```
Comment (chejennifer) [inline static/js/chart/draw_bar.ts:703]: nit: it looks like this same mapping is done multiple times in this file.. could we make a helper function for this?

**Review by chejennifer — COMMENTED** [review]
nice!

File: static/js/chart/draw_utils.ts [hunk 2]
```diff
@@ -47,22 +46,23 @@ export const MAX_UNIT_LENGTH = 5;
...
+): d3.Selection<HTMLDivElement, any, any, any> {
+  const tooltipRef = _.uniqueId(`${TOOLTIP_ID}-`);
   container
     .attr("style", "position: relative")
     .append("div")
-    .attr("id", TOOLTIP_ID)
+    .attr("id", tooltipRef)
+    .attr("class", "draw-tooltip")
     .attr("style", "position: absolute; display: none; z-index: 1");
+  return container.select(`#${tooltipRef}`);
```
[the remark this answers is not in the document]
Comment (juliawu) [reply static/js/chart/draw_utils.ts:66]: Good call. Added.

File: static/js/chart/draw_bar.ts [hunk 3]
```diff
@@ -51,6 +53,144 @@ export const HORIZONTAL_BAR_CHART = {
...
+
+  // define tooltip mouse behavior
+  const mouseoverFn = function () {
+    if (hideFn) {
+      clearTimeout(hideFn);
+    }
+    tooltip.style("display", "block");
+    const tooltipContent = getTooltipContent(d3.select(this));
+    tooltip.html(tooltipContent);
+    d3.select(this).style("opacity", 0.5);
```
[the remark this answers is not in the document]
Comment (juliawu) [reply static/js/chart/draw_bar.ts]: This is a great idea. Edited so now only the highlighted bar has higher opacity.

**Review by chejennifer — APPROVED** [review]
nice, thanks for the updates!

**Pull request merged**